File: apps/android/app/src/main/java/fit/aquazero/app/core/common/MealTrust.kt
package fit.aquazero.app.core.common

/** Match confidence band for confirm-first meal UI. */
enum class MatchConfidenceBand {
    HIGH,
    MODERATE,
    LOW,
}

object MealTrust {

    const val FAT_CAUTION_NOTE =
        "Photo and AI estimates often miss cooking oil and sauces. Add fat if needed and confirm portions."

    fun confidenceBandFromScore(score: Int): MatchConfidenceBand = when {
        score >= 80 -> MatchConfidenceBand.HIGH
        score >= 55 -> MatchConfidenceBand.MODERATE
        else -> MatchConfidenceBand.LOW
    }

    fun confidenceBandLabel(band: MatchConfidenceBand): String = when (band) {
        MatchConfidenceBand.HIGH -> "High match"
        MatchConfidenceBand.MODERATE -> "Moderate match"
        MatchConfidenceBand.LOW -> "Low match"
    }

    fun shouldShowFatCaution(itemNames: List<String>, kcal: Double, fatG: Double): Boolean {
        val joined = itemNames.joinToString(" ").lowercase()
        val cue = Regex(
            """\b(fried|deep.?fried|crispy|sauce|gravy|dressing|mayo|butter|oil|cheese|""" +
                """cream|bacon|sausage|curry|stir.?fry|tempura|saut[eé])\b""",
        ).containsMatchIn(joined)
        val fatShare = if (kcal > 0) fatG * 9 / kcal else 0.0
        return cue || fatShare >= 0.45
    }

    data class CookingFatPreset(
        val id: String,
        val label: String,
        val grams: Double,
        val kcal: Double,
        val fatG: Double,
    )

    val cookingFatPresets: List<CookingFatPreset> = listOf(
        CookingFatPreset("tsp-oil", "1 tsp oil", 5.0, 45.0, 5.0),
        CookingFatPreset("tbsp-oil", "1 tbsp oil", 14.0, 126.0, 14.0),
        CookingFatPreset("tbsp-sauce", "1 tbsp sauce", 15.0, 30.0, 2.0),
    )

    fun portionCorrectionWorthRemembering(defaultGrams: Int, confirmedGrams: Int): Boolean {
        if (defaultGrams <= 0 || confirmedGrams <= 0) return false
        val ratio = confirmedGrams.toDouble() / defaultGrams
        return ratio < 0.8 || ratio > 1.2
    }
}

File: packages/shared/src/mealTrust.ts
/**
 * Confirm-first trust helpers shared across web and API tests.
 * Keep in sync with apps/api/src/modules/chat/mealTrust.ts.
 */

export type MatchConfidenceBand = 'high' | 'moderate' | 'low';

export function confidenceBandFromScore(score: number): MatchConfidenceBand {
  if (score >= 80) return 'high';
  if (score >= 55) return 'moderate';
  return 'low';
}

export function confidenceBandLabel(band: MatchConfidenceBand): string {
  if (band === 'high') return 'High match';
  if (band === 'moderate') return 'Moderate match';
  return 'Low match';
}

export function shouldShowFatCaution(
  itemNames: readonly string[],
  totals: { kcal: number; fatG: number },
): boolean {
  const joined = itemNames.join(' ').toLowerCase();
  const cue =
    /\b(fried|deep.?fried|crispy|sauce|gravy|dressing|mayo|butter|oil|cheese|cream|bacon|sausage|curry|stir.?fry|tempura|saut[eé])\b/.test(
      joined,
    );
  const fatShare = totals.kcal > 0 ? (totals.fatG * 9) / totals.kcal : 0;
  return cue || fatShare >= 0.45;
}

export const FAT_CAUTION_NOTE =
  'Photo and AI estimates often miss cooking oil and sauces. Add fat if needed and confirm portions.';

export interface CookingFatPreset {
  id: string;
  label: string;
  grams: number;
  kcal: number;
  fatG: number;
}

export const COOKING_FAT_PRESETS: readonly CookingFatPreset[] = [
  { id: 'tsp-oil', label: '1 tsp oil', grams: 5, kcal: 45, fatG: 5 },
  { id: 'tbsp-oil', label: '1 tbsp oil', grams: 14, kcal: 126, fatG: 14 },
  { id: 'tbsp-sauce', label: '1 tbsp sauce', grams: 15, kcal: 30, fatG: 2 },
] as const;

export function portionCorrectionWorthRemembering(defaultGrams: number, confirmedGrams: number): boolean {
  if (defaultGrams <= 0 || confirmedGrams <= 0) return false;
  const ratio = confirmedGrams / defaultGrams;
  return ratio < 0.8 || ratio > 1.2;
}
